### PR TITLE
Optimizer settings in HERON input

### LINKED ISSUE #132: [TASK] Enable value at risk in template
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
The statistical metric used as the objective for optimization is currently expected value. The user should be able to specify another metric such as value at risk.

**Describe the solution you'd like**
Allow the user to specify in the input file which statistical metric will be used for the optimization objective. If not given, it should fall back to expected value.

**Describe alternatives you've considered**
none

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [ ] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by dgarrett622**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes #132 

##### What are the significant changes in functionality due to this change request?

- EconomicRatio PostProcessor now used (BasicStatistics metrics available through EconomicRatio)
- User can specify statistical metric used as objective for optimization in the HERON input file
- User can specify min or max for optimization in the HERON input file
- Uploaded integration test for new functionality
- Updated Tester so that HERON integration tests can be run locally on Windows

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [x] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: src/Cases.py [hunk 1]
```diff
@@ -182,6 +200,53 @@ def get_input_specs(cls):
...
+                                \default{`zero'}
+                                \item requested $ \alpha $ value (a floating point value between 0.0
+                                and 1.0). Required when \xmlNode{metric} is ``expectedShortfall'' or
+                                ``valueAtRisk.'' \default{5.0}
+                              \end{itemize}""")
+    optimizer.addSub(metric)
+    type_options = InputTypes.makeEnumType('TypeOptions', 'TypeOptionsType',
+                                           ['min', 'max'])
+    desc_type_options = r"""determines whether the objective should be minimized or maximized.
+                            \default{max}"""
```
Comment (PaulTalbot-INL) [inline src/Cases.py]: How painful would it be to have the default based on the metric? I suspect we want default for VaR to be minimize, for example.

File: src/Testers/HeronIntegrationTester.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def __init__(self, name, param):
       @ Out, None
     """
     RavenTester.__init__(self, name, param)
-    self.heron_driver = os.path.join(HERON_LOC, '..', 'heron')
+    self.heron_driver = os.path.join(HERON_LOC, 'main.py')
```
Comment (PaulTalbot-INL) [inline src/Testers/HeronIntegrationTester.py]: We should talk about this change; does the "heron" executable not work on Windows at all?

The "executable" lets us handle things like loading the conda env, so I'm not thrilled by bypassing it entirely in the testing system.

File: tests/integration_tests/mechanics/optimization_settings/heron_input.xml [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+          </reference_price>
+        </CashFlow>
+      </economics>
+    </Component>
+
+    <!-- design notes
+    add a <resource> node for at least CashFlow drivers
+    add multiplier and absolute value options for manipulating entries
+    maybe disallow <linear> and <resource> except for applicable ValuedParams?
+    -->
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/mechanics/optimization_settings/heron_input.xml]: can we remove the un-useful commented blocks in this test? I know they need to be removed from some of the other tests as well.

File: tests/integration_tests/mechanics/optimization_settings/write_inner.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import xml.etree.ElementTree as ET
```
Comment (PaulTalbot-INL) [inline tests/integration_tests/mechanics/optimization_settings/write_inner.py]: this file is generated by the heron template run, and so doesn't need to be included in the repo

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
A couple minor comments to consider

File: src/Cases.py (hunk 1 above)
Comment (dgarrett622) [reply src/Cases.py]: I could set up a default mapping that maps each metric to min or max. This should not be difficult to implement.

File: tests/integration_tests/mechanics/optimization_settings/heron_input.xml (hunk 3 above)
Comment (dgarrett622) [reply tests/integration_tests/mechanics/optimization_settings/heron_input.xml]: I can clean up the comments in this test. Should I remove the commented blocks in other tests for this PR?

File: tests/integration_tests/mechanics/optimization_settings/write_inner.py (hunk 4 above)
Comment (dgarrett622) [reply tests/integration_tests/mechanics/optimization_settings/write_inner.py]: Oops, grabbed one too many files. I'll remove it.

File: tests/integration_tests/mechanics/optimization_settings/heron_input.xml (hunk 3 above)
Comment (PaulTalbot-INL) [reply tests/integration_tests/mechanics/optimization_settings/heron_input.xml]: I think just this one is fine, but if you want to clear out the others I would not object!

File: src/Testers/HeronIntegrationTester.py (hunk 2 above)
Comment (dgarrett622) [reply src/Testers/HeronIntegrationTester.py]: Let's talk about this in our developer meeting.

Comment (dgarrett622) [reply src/Testers/HeronIntegrationTester.py]: Looks like the original implementation does not work on my Windows setup. However, prepending "bash.exe" to the "heron" executable seems to do the trick. It looks like that's how RrR works as well. I'll implement a check for Windows OS and add "bash.exe" to keep the "heron" executable.

Comment (PaulTalbot-INL) [reply src/Testers/HeronIntegrationTester.py]: It's good to have a dev on windows! We'll find a lot more of these issues a priori.

**Review by PaulTalbot-INL — APPROVED** [review]

**Pull request merged**